### PR TITLE
Farcaster subgraph for airstack social schema

### PR DESCRIPTION
**Pull request by rahul7668gupta**
This PR marks the completion of farcaster subgraph for airstack social schema.

Pending:
- [x] unit tests

Tests screenshot
Screenshot 2023-02-09 at 2 24 56 PM

### CODE REVIEW

**Comment by rahul7668gupta** [comment]
merge this PR only after ens subgraph is merged!

File: airstack-modules/modules/airstack/domain-name/Readme.md
Comment (sharathkrml) [inline airstack-modules/modules/airstack/domain-name/Readme.md:6]:
```suggestion
    block: ethereum.Block,            #ethereum block object in subgraph
```
remove duplication

File: ens/modules/airstack/domain-name/Readme.md
Comment (sharathkrml) [inline ens/modules/airstack/domain-name/Readme.md:6]:
```suggestion
    block: ethereum.Block,            #ethereum block object in subgraph 
```
remove duplication

**Review by sharathkrml — COMMENTED** [review]
one small change requested

File: ens/modules/airstack/domain-name/Readme.md
Comment (rahul7668gupta) [reply ens/modules/airstack/domain-name/Readme.md]: fixed

**Comment by rahul7668gupta** [comment]
requesting re review

**Review by sharathkrml — COMMENTED** [review]
looks fine, @0xsarvesh please have a look

**Comment by rahul7668gupta** [comment]
added tokenId in profile and user reg txn
Screenshot 2023-02-09 at 11 17 46 PM

File: airstack-modules/modules/airstack/common/index.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/common/index.ts:84]: Instead of returning `unknown` it should throw an error. Similarly for `getNetwork`

File: airstack-modules/modules/airstack/common/index.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/common/index.ts:178]: **Suggestion:** If we can create an empty object with only id. Then the caller can easily identify by comparing some of the fields, if its new object or old object. It will be very easy to write few logics on top of that.

The assigning of the values can be done by the caller.

This is just a suggestion, not expecting to change anything as of now here.

File: farcaster/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline farcaster/modules/airstack/social/social.ts:32]: We can have the above 3 fields in one as `block`

File: farcaster/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline farcaster/modules/airstack/social/social.ts:268]:
```suggestion
  export class AirExtraData {
```

File: farcaster/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline farcaster/modules/airstack/social/social.ts:47]: The key generation logic should be inside `getOrCreateAirUser` or expose a new function.

File: farcaster/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline farcaster/modules/airstack/social/social.ts:119]: Not using from common?

File: farcaster/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline farcaster/modules/airstack/social/social.ts:188]: add `tokenAdress`

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
Added few comments.

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:83]: done

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:53]: done

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:57]: done

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:62]: yes it can happen later on when the schema changes, done

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:59]: done

File: airstack-modules/modules/airstack/common/index.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/common/index.ts:178]: ok

File: farcaster/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply farcaster/modules/airstack/social/social.ts:47]: moved inside, but is required here for air extras, done

File: farcaster/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply farcaster/modules/airstack/social/social.ts:119]: not sure if i understand

File: farcaster/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply farcaster/modules/airstack/social/social.ts:268]: done

File: farcaster/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply farcaster/modules/airstack/social/social.ts:188]: done

File: airstack-modules/modules/airstack/common/index.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/common/index.ts]: throwing error now

File: farcaster/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply farcaster/modules/airstack/social/social.ts]: done

File: farcaster/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply farcaster/modules/airstack/social/social.ts]: done

File: airstack-modules/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/social/social.ts:235]: add {} in if statements.

File: airstack-modules/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/social/social.ts:149]: This is a create function so it can save the entity.

File: airstack-modules/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/social/social.ts:187]: This is a create function, it can save the entity.

File: airstack-modules/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/social/social.ts:179]: Add { } explicitly.

File: airstack-modules/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/social/social.ts:138]: This is getOrCreate function so the saving of entity should be outside.
May be pass the airAccountId as param. Or you can handle this outside.

File: airstack-modules/modules/airstack/social/social.ts
Comment (deepesh-kn) [inline airstack-modules/modules/airstack/social/social.ts:74]: You already have the  for loop you can create the `airUserExtraIds` array in the loop itself.

**Review by deepesh-kn — CHANGES_REQUESTED** [review]

File: airstack-modules/modules/airstack/common/index.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/common/index.ts:87]: Thesis because compiler throws error even after null check is done.
Screenshot 2023-02-16 at 11 13 58 PM

File: airstack-modules/modules/airstack/common/index.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/common/index.ts:78]: same as above

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:20]: done

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:46]: done

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:235]: done

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:179]: doen

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:138]: it is being done for air user. but if the air user does not exist, air account is being created an saved outside the `getOrCreateAirAccount` func.

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:74]: done

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:187]: done

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:149]: renamed to `getorCreateAirProfile`

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:39]: created classes for user profile and transfer

**Comment by rahul7668gupta** [comment]
@deepesh-kn this is up for re-review.

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:120]: from and to fields from AirTransaction already do track the same data, where from is the previous owner and to is the new owner. Please confirm if we still need to add them.

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:130]: from and to fields from AirTransaction already do track the same data, where from is the previous owner and to is the new owner. Please confirm if we still need to add them.

File: airstack-modules/graphql/airstack-nft-marketplace-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-nft-marketplace-schema.graphql.ts]: fixed

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts]: fixed

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]

File: farcaster/src/utils.ts
Comment (rahul7668gupta) [reply farcaster/src/utils.ts:78]: done

File: airstack-modules/graphql/airstack-social-schema.graphql.ts
Comment (rahul7668gupta) [reply airstack-modules/graphql/airstack-social-schema.graphql.ts:52]: done

File: farcaster/src/mapping.ts
Comment (rahul7668gupta) [reply farcaster/src/mapping.ts:71]: done

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:59]: doen

File: airstack-modules/modules/airstack/social/social.ts
Comment (rahul7668gupta) [reply airstack-modules/modules/airstack/social/social.ts:73]: done

**Review by 0xsarvesh — APPROVED** [review]

**Review by deepesh-kn — APPROVED** [review]

**Review by sharathkrml — APPROVED** [review]

**Pull request merged**
